runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MetricResultsProtos.java

```diff
@@ -0,0 +1,235 @@
+        BeamFnApi.MetricResults.Builder::addAttempted);
+    try {
+      process(
+          builder,
+          metricKey,
+          metricResult.getCommitted(),
+          set,
+          BeamFnApi.MetricResults.Builder::addCommitted);
+    } catch (UnsupportedOperationException ignored) {
+    }
+  }
+
+  /**
+   * Add a metric key and value (from a {@link MetricResult}) to a {@link
+   * BeamFnApi.MetricResults.Builder}.
+   *
+   * @param builder proto {@link BeamFnApi.MetricResults.Builder} to add to
+   * @param set set this metric's value on a {@link SimpleMonitoringInfoBuilder} (via an API that's
+   *     specific to this metric's type, as provided by the caller)
+   * @param add add the {@link MonitoringInfo} created from this key/value to either the "attempted"
+   *     or "committed" list of the {@link BeamFnApi.MetricResults.Builder builder}
+   */
+  private static <T> void process(
+      BeamFnApi.MetricResults.Builder builder,
+      MetricKey metricKey,
+      T value,
+      BiConsumer<SimpleMonitoringInfoBuilder, T> set,
+      BiConsumer<BeamFnApi.MetricResults.Builder, MonitoringInfo> add) {
+    if (value != null) {
+      SimpleMonitoringInfoBuilder partial =
+          new SimpleMonitoringInfoBuilder().handleMetricKey(metricKey);
+      ;
+      set.accept(partial, value);
+      MonitoringInfo monitoringInfo = partial.build();
+      if (monitoringInfo != null) {
+        add.accept(builder, monitoringInfo);
+      }
+    }
+  }
+
+  // `fromProto` helpers
+
+  /**
+   * Helper for converting {@link BeamFnApi.MetricResults} to {@link MetricResults}.
+   *
+   * <p>The former separates "attempted" and "committed" metrics at the highest level, while the
+   * latter splits on metric-type (counter, distribution, or gauge) first, so converting basically
+   * amounts pivoting those between those two.
+   */
+  private static class MetricResultsBuilder {
+    private final Map<MetricKey, MetricResultBuilder<Long>> counters = new ConcurrentHashMap<>();
+    private final Map<MetricKey, MetricResultBuilder<DistributionResult>> distributions =
+        new ConcurrentHashMap<>();
+    private final Map<MetricKey, MetricResultBuilder<GaugeResult>> gauges =
+        new ConcurrentHashMap<>();
+
+    /**
+     * Populate metric-type-specific maps with {@link MonitoringInfo}s from a {@link
+     * BeamFnApi.MetricResults}.
+     */
+    public MetricResultsBuilder(BeamFnApi.MetricResults metrics) {
+      add(metrics.getAttemptedList(), false);
+      add(metrics.getCommittedList(), true);
+    }
+
+    private void add(Iterable<MonitoringInfo> monitoringInfos, Boolean committed) {
+      for (MonitoringInfo monitoringInfo : monitoringInfos) {
+        add(monitoringInfo, committed);
+      }
+    }
+
+    public MetricResults build() {
+      return new DefaultMetricResults(build(counters), build(distributions), build(gauges));
+    }
+
+    /** Helper for turning a map of result-builders into a sequence of results. */
```

Comment:
Please remove helpers if they aren't called in more than one place. I have a feeling some of these add( calls might be in this situation